### PR TITLE
WT-8365 Improve performance for timestamp_transaction

### PR DESCRIPTION
**Pull request by ddanderson**
See the ticket for the rationale and benefits of using this approach to gain performance.

### CODE REVIEW

File: src/txn/txn_timestamp.c [hunk 1]
```diff
@@ -854,50 +854,67 @@ __wt_txn_set_read_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t read_ts)
...
+    WT_ASSERT(session, cfg[0] != NULL && cfg[1] != NULL && cfg[2] == NULL);
+    __wt_config_init(session, &cparser, cfg[1]);
+    while ((ret = __wt_config_next(&cparser, &ckey, &cval)) == 0) {
+        WT_ASSERT(session, ckey.str != NULL);
+        if (WT_STRING_MATCH("commit_timestamp", ckey.str, ckey.len)) {
+            WT_RET(__wt_txn_parse_timestamp(session, "commit", &commit_ts, &cval));
+            set_ts = true;
+        } else if (WT_STRING_MATCH("durable_timestamp", ckey.str, ckey.len)) {
+            WT_RET(__wt_txn_parse_timestamp(session, "durable", &durable_ts, &cval));
+            set_ts = true;
```
Comment (ddanderson) [inline src/txn/txn_timestamp.c:887]: I want to call out the setting of the `set_ts` variable when `durable_timestamp` is specified and also when `prepare_timestamp` is specified.  The original code does not set this for these cases, and that seems incorrect to me, and so is "fixed" here.  As I understand it, having this be true buts an informational message in the log (Added in WT-4712).  So we could potentially be missing some of these informational messages.  Though it's possible MongoDB never triggers this problem depending on how it uses the API.

@sueloverso, you made the original change, please let me know if I misunderstand this.

File: src/txn/txn_timestamp.c [hunk 2]
```diff
@@ -854,50 +854,67 @@ __wt_txn_set_read_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t read_ts)
...
-    WT_RET(__wt_config_gets_def(session, cfg, "prepare_timestamp", 0, &cval));
-    if (ret == 0 && cval.len != 0) {
-        WT_RET(__wt_txn_parse_timestamp(session, "prepare", &ts, &cval));
-        WT_RET(__wt_txn_set_prepare_timestamp(session, ts));
-    }
+    if (prepare_ts != 0)
+        WT_RET(__wt_txn_set_prepare_timestamp(session, prepare_ts));
+
     if (set_ts)
         WT_RET(__wt_txn_ts_log(session));
```
Comment (sueloverso) [inline src/txn/txn_timestamp.c:919]: So related to the fix you did, can you update one or more of the `test_debug_mode` python tests to set/use durable and prepare timestamps and check the output.

**Review by sueloverso — APPROVED** [review]
LGTM with a couple minor things.

**Review by sulabhM — APPROVED** [review]

File: src/txn/txn_timestamp.c (hunk 2 above)
Comment (ddanderson) [reply src/txn/txn_timestamp.c:920]: Added a test to check that this is working.

File: src/txn/txn_timestamp.c [hunk 3]
```diff
@@ -854,50 +854,67 @@ __wt_txn_set_read_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t read_ts)
...
 
-    /* Look for a commit timestamp. */
-    ret = __wt_config_gets_def(session, cfg, "commit_timestamp", 0, &cval);
-    WT_RET_NOTFOUND_OK(ret);
-    if (ret == 0 && cval.len != 0) {
-        WT_RET(__wt_txn_parse_timestamp(session, "commit", &ts, &cval));
-        WT_RET(__wt_txn_set_commit_timestamp(session, ts));
-        set_ts = true;
+    /*
+     * If the API received no configuration string, there's nothing to do.
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply src/txn/txn_timestamp.c]: I expanded this comment to make this clearer.

**Comment by ddanderson** [comment]
@sueloverso, I just tagged you to re-review as I added a python test for this.

**Review by sueloverso — APPROVED** [review]
LGTM

**Pull request merged**